segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java

```diff
@@ -0,0 +1,120 @@
+        long offset = applySearch(store, segment, targetOffset, greater).getValue();
+        return offset;
+    }
+
+    /**
+     * Locate the requested offset in index segment.
+     *
+     * @param store  The StreamSegmentStore to attach to (and issue requests to).
+     * @param segment Segment name.
+     * @param targetOffset The requested offset's corresponding position to search in the index segment entry.
+     * @param greater boolean to determine if the next higher or the lower value to be returned in case the requested offset is not present.
+     *
+     * @return the corresponding offset of index segment.
+     */
+    public static long locateOffsetForIndexSegment(StreamSegmentStore store, String segment, long targetOffset, boolean greater) {
+        return applySearch(store, segment, targetOffset, greater).getKey() * NameUtils.INDEX_APPEND_EVENT_SIZE;
+    }
+
+    private static Map.Entry<Long, Long> applySearch(StreamSegmentStore store, String segment, long targetOffset, boolean greater) {
+        String indexSegmentName = NameUtils.getIndexSegmentName(segment);
+
+        //Fetch start and end idx.
+        SegmentProperties properties = store.getStreamSegmentInfo(indexSegmentName, TIMEOUT).join();
+        long startIdx = properties.getStartOffset() / NameUtils.INDEX_APPEND_EVENT_SIZE;
+        long endIdx = properties.getLength() / NameUtils.INDEX_APPEND_EVENT_SIZE;
+        //If startIdx and endIdx are same, then pass length of segment as a result.
+        if (startIdx == endIdx) {
+           return getSegmentLength(store, segment, startIdx);
+        }
+
+        return SortUtils.newtonianSearch(idx -> {
+            ReadResult result = store.read(indexSegmentName, idx * NameUtils.INDEX_APPEND_EVENT_SIZE, NameUtils.INDEX_APPEND_EVENT_SIZE, TIMEOUT).join();
+            ReadResultEntry firstElement = result.next();
+            // TODO deal with element which is split over multiple entries.
+            switch (firstElement.getType()) {
+                case Cache: // fallthrough
+                case Storage:
+                    return getOffsetFromIndexEntry(firstElement);
+                case Future:
+                    return RETRY_WITH_BACKOFF.retryWhen(e -> true).run(() -> getOffsetFromIndexEntry(firstElement));
+                case Truncated:
+                    throw new SegmentTruncatedException(String.format("Segment %s has been truncated.", segment));
+                case EndOfStreamSegment:
+                    throw new IllegalStateException(String.format("Unexpected size of index segment of type: %s was encountered for segment %s.", firstElement.getType(), segment));
+            }
+            throw new IllegalStateException(String.format("Unexpected index segment of type: %s was encountered for segment %s.", firstElement.getType(), segment));
+        }, startIdx, endIdx > 0 ? endIdx - 1 : 0, targetOffset, greater);
```

Comment:
Do we still need this conditional check here `endIdx > 0 ? endIdx - 1 : 0` . Since we have already handled the empty index segment check at line 87.